pkg/kv/kvclient/kvstreamer/streamer.go

```diff
@@ -391,13 +393,22 @@ func (s *Streamer) Enqueue(
 	// TODO(yuzefovich): we might want to have more fine-grained lock
 	// acquisitions once pipelining is implemented.
 	s.mu.Lock()
+	locked := true
 	defer func() {
-		if retErr != nil && s.mu.err == nil {
-			// Set the error so that mainLoop of the worker coordinator exits
-			// as soon as possible, without issuing any more requests.
-			s.mu.err = retErr
+		if retErr != nil {
+			if !locked {
+				s.mu.Lock()
+				locked = true
```

Comment:
This is introducing a new race condition, which is objectively worse - it can only be caught by the race detector now, which is less precise/powerful than the deadlock detector.

After `s.mu.Unlock()` in the main body of the function, there's a brief amount of time where `locked` is still `true`.
If the defer is run at that moment, then this condition here `if !locked` will not fire, and `s.mu.err` below becomes a racy write.

Can't this be handled with a buffered channel for s.mu.err instead?

Reply:
Indeed, nice catch, thanks! I opened up a fix in #75078.

> Can't this be handled with a buffered channel for s.mu.err instead?

The buffered channel could work, but it doesn't seem to be a natural fit for this since, generally speaking, we want to observe the error from multiple goroutines (`workerCoordinator.mainLoop` and `GetResults`).